src/ucp/api/ucp.h

```diff
@@ -494,6 +494,24 @@ typedef enum {
 } ucp_stream_recv_flags_t;
 
 
+/**
+ * @ingroup UCP_COMM
+ * @brief UCP operation fields and flags
+ *
+ * The enumeration allows specifying which fields in @ref ucp_op_info_t are
+ * present and operation flags are used. It is used to enable backward
+ * compatibility support.
+ */
+enum ucp_op_attr {
+    UCP_OP_ATTR_FIELD_REQUEST    = UCS_BIT(0), /**< request field */
+    UCP_OP_ATTR_FIELD_CB         = UCS_BIT(1), /**< cb field */
+    UCP_OP_ATTR_FIELD_USER_DATA  = UCS_BIT(2), /**< user_data field */
+    UCP_OP_ATTR_FIELD_DATATYPE   = UCS_BIT(3), /**< datatype field */
+
+    UCP_OP_ATTR_FLAG_NO_IMM_CMPL = UCS_BIT(16) /**< deny immediate completion */
+};
+
+
```

Comment:
@yosefe @shamisp @Akshay-Venkatesh  with this API change, can we take the opportunity to define memory types at UCP api level and pass memory type as an attribute if users know the memory type?  to avoid memory detection in the UCX library?

Reply:
@bureddy yes, it should be possible to add another flag and field to op_info to indicate memory type (or even pass memory handle)
the intention of op_info is to provide extensions for non-fast-path cases without introducing extra arguments or breaking backward compatibility.

Reply:
@yosefe this means we will have additional branches on the UCP level to dispatch the options, which will impact the fast path as well ?

Reply:
@shamisp the idea is to hide the new branches below existing slow-path branches. for example. check memory type flag only if we have GPU devices, check memory handle only for zero-copy path, check request only for non-immediate completion, etc.

Reply:
Conceptually makes sense. It would help to take a look at the actual implementation to see what is the overall impact. Does not have to be final.